test/test_xbuffer.cpp

```diff
@@ -0,0 +1,82 @@
+/***************************************************************************
+* Copyright (c) QuantStack                                                 *
+*                                                                          *
+* Distributed under the terms of the BSD 3-Clause License.                 *
+*                                                                          *
+* The full license is in the file LICENSE, distributed with this software. *
+***************************************************************************/
+
+#include "doctest/doctest.h"
+
+#include <numeric>
+
+#include "xparrow/xbuffer.hpp"
+
+namespace xp
+{
+    using buffer_test_type = xbuffer<uint8_t>;
+
+    namespace
+    {
+        uint8_t* make_test_buffer(std::size_t size, uint8_t start_value = 0)
+        {
+            uint8_t* res = new uint8_t[size];
+            std::iota(res, res + size, start_value);
+            return res;
+        }
+    }
+
+    TEST_SUITE("xbuffer")
+    {
+        TEST_CASE("constructors")
+        {
+            buffer_test_type b0;
+            CHECK_EQ(b0.data(), nullptr);
+            CHECK_EQ(b0.size(), 0u);
+
+            std::size_t size = 4u;
+            buffer_test_type b1(size);
+            CHECK_NE(b1.data(), nullptr);
+            CHECK_EQ(b1.size(), size);
+
+            uint8_t* mem = make_test_buffer(size);
+            buffer_test_type b2(mem, size);
+            CHECK_EQ(b2.data(), mem);
+            CHECK_EQ(b2.size(), size);
+            CHECK_EQ(b2.data()[2], uint8_t(2));
+        }
+
+        TEST_CASE("copy semantic")
+        {
+            std::size_t size = 4;
+            buffer_test_type b1(make_test_buffer(size), size);
+            buffer_test_type b2(b1);
+            CHECK_EQ(b1, b2);
+
+            std::size_t size2 = 8;
+            buffer_test_type b3(make_test_buffer(size2, 4), size2);
+            b2 = b3;
+            CHECK_EQ(b2, b3);
+            CHECK_NE(b1, b2);
+        }
+
+        TEST_CASE("move semantic")
+        {
+            std::size_t size = 4;
+            buffer_test_type b1(make_test_buffer(size), size);
+            buffer_test_type control(b1);
+            buffer_test_type b2(std::move(b1));
+            CHECK_EQ(b2, control);
+            CHECK_EQ(b1.size(), 0);
```

Comment:
Note that if `b2` was not empty, then at this point `b1` would not be empty. I'm thinking we might want to not swap in the move operations and release immediately  (on move) the moved-from object internals. Any reason against?

Reply:
I find the implementation with `swap` quite elegant, otherwise we would have to:
- either deallocating the buffer in the destination, assign the pointer from source, reset the pointer in source
- or swap, and then immediatly delete in the source.

That's more code for "almost" the same result (we don't really care that the memory is not freed immediatly).

The test is misleading here, the only think I want to check is that `b1` did not keep a copy of the pointer, leading to a double destruction. And I want to check this with an fialure rather than a crash (only).

Reply:
I just worry that having data in the moved-from object will lead to false-positive situations when reading them.

Well as long as the size says zero I guess it's ok, accessing the values is undefined-behavior (we might want to specify the contract although it's kind of obvious).